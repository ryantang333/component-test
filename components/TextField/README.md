# @test/text-field

Welcome to my `TextField` component.

## Installation

```sh
npm i @test/text-field
# or with yarn
yarn add @test/text-field
```

Then to use the component in your code just import it!

```js
import TextField from '@test/text-field';
// and with css-modules
import '@test/text-field/dist/main.css';
```
