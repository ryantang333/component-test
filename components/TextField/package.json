{
  "name": "@test/text-field",
  "version": "0.0.0",
  "main": "./dist/cjs/index.js",
  "module": "./dist/esm/index.js",
  "types": "./dist",
  "repository": "https://github.com/ryantang333/component-test",
  "author": "ryan-tang ryan_tang@apple.com",
  "contributors": [
    "ryan-tang ryan_tang@apple.com"
  ],
  "license": "MIT",
  "sideEffects": [
    "*.css"
  ],
  "scripts": {
    "clean": "ds clean",
    "build": "ds build",
    "dev": "ds dev",
    "storybook": "ds storybook start",
    "start": "ds build --watch",
    "test": "ds test",
    "lint": "ds lint",
    "size": "ds size"
  },
  "dependencies": {
    "@babel/runtime": "7.11.2"
  },
  "peerDependencies": {
    "react": ">= 16.8.6"
  },
  "files": [
    "dist",
    "src",
    "!*.snippet.*",
    "!*theme.*",
    "!*.test.*",
    "!__snapshots__",
    "!__tests__"
  ]
}
