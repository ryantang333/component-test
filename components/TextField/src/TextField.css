.text-field {
  border: 1px solid #000;
  border-radius: 4px;
  position: relative;
}

.text-field.disabled {
  border: 1px solid rgba(0, 0, 0, 0.38);
}

.helper-text {
  color: rgba(0, 0, 0, 0.6);
  font-family: 'Roboto', 'Helvetica', 'Arial', sans-serif;
  font-weight: 400;
  font-size: 0.75rem;
  text-align: left;
  margin-top: 0.1875rem;
  margin-right: 0.875rem;
  margin-bottom: 0;
  margin-left: 0.875rem;
}

.text-field.error {
  border: 2px groove #d32f2f;
}

.text-field label.error,
.helper-text.error {
  /* would use sass variables or something similar for these colors */
  color: #d32f2f;
}
