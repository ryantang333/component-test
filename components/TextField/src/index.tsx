import React from 'react';
import cn from 'classnames';
import InputField from '../../InputField/src';
import InputLabel from '../../InputLabel/src';
import style from './TextField.css';

interface TextFieldProps {
  /** User-provided components, each key corresponds to an associated inner component */
  components?: object;
  /** Default value */
  defaultValue?: string;
  /** Disabled */
  disabled?: boolean;
  /** Error boolean to change styles */
  error?: boolean;
  /** Helper text below the input */
  helperText?: string;
  /** ID for input and corresponding label */
  id?: string;
  /** Input label */
  label?: string;
  /** The onChange function */
  onChange?: React.ChangeEventHandler<HTMLInputElement>;
  /** Read only */
  readOnly?: boolean;
  /** Marks input as required (adds an asterisk) */
  required?: boolean;
  /** User-provided styles, each key corresponds to an associated inner component */
  styles?: object;
  /** Input type */
  type?: string;
}

/** A simple component. */
export const TextField = (props: TextFieldProps) => {
  const {
    components = {},
    defaultValue,
    disabled,
    error,
    helperText,
    id,
    onChange,
    readOnly,
    required,
    styles = {},
    type = 'text'
  } = props;
  const classNames = cn(
    style.textField,
    { [style.disabled]: disabled },
    { [style.error]: error }
  );
  const helperTextClasses = cn(style.helperText, { [style.error]: error });

  // user-provided custom components
  const InputFieldComponent = components?.InputField || InputField;
  const InputLabelComponent = components?.InputLabel || InputLabel;

  // user-provided custom styles
  const InputFieldStyles = styles?.InputField;
  const InputLabelStyles = styles?.InputLabel;

  let { label } = props;

  if (label && required) {
    label += ' *';
  }

  return (
    <>
      <div className={classNames}>
        {label && (
          <InputLabelComponent
            error={error}
            htmlFor={id}
            label={label}
            style={InputLabelStyles}
          />
        )}
        <InputFieldComponent
          defaultValue={defaultValue}
          disabled={disabled}
          id={id}
          readOnly={readOnly}
          required={required}
          style={InputFieldStyles}
          type={type}
          onChange={onChange}
        />
      </div>
      <p className={helperTextClasses}>{helperText}</p>
    </>
  );
};

export default TextField;
