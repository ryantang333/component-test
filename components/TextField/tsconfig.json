{
  "extends": "../../tsconfig.build.json",
  "include": ["src/**/*", "../../typings/**/*", "../InputLabel"],

  "compilerOptions": {
    "outDir": "./dist",
    "rootDir": "./src",
    "composite": true,

    // Need to override emotion's types because they pollute the global
    "baseUrl": ".",
    "paths": {
      "@emotion/core": ["../../typings/emotion.d.ts"]
    }
  }
}
