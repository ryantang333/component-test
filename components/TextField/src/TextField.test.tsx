import React from 'react';
import { fireEvent, render, screen } from '@testing-library/react';
import TextField from '.';

describe('TextField', () => {
  function renderComponent(props?: any) {
    return render(<TextField {...props} />);
  }

  test('It should have a default value', () => {
    const defaultValue = 'Default Value';

    renderComponent({
      defaultValue
    });

    expect(screen.getByDisplayValue(defaultValue)).toBeInTheDocument();
  });

  test('It should not have a default value', () => {
    const defaultValue = 'Default Value';

    renderComponent();

    expect(screen.queryByDisplayValue(defaultValue)).toBeNull();
  });

  test('It should be disabled', () => {
    const defaultValue = 'Default Value';

    renderComponent({
      defaultValue,
      disabled: true
    });

    expect(screen.getByDisplayValue(defaultValue)).toBeDisabled();
  });

  test('It calls on change', () => {
    const defaultValue = 'Default Value';
    const mockChange = jest.fn();

    renderComponent({
      defaultValue,
      disabled: true,
      onChange: mockChange
    });

    const input = screen.getByDisplayValue(defaultValue);

    fireEvent.change(input, { target: { value: 'a' } });
    expect(mockChange).toHaveBeenCalled();
  });

  test('It matches the snapshot', () => {
    const { container } = render(<TextField />);
    expect(container).toMatchSnapshot();
  });
});
