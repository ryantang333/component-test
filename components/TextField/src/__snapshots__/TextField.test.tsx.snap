// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextField It calls on change 1`] = `
<div>
  <div
    class="textField"
  >
    <input
      class="inputField"
      type="text"
      value=""
    />
  </div>
  <p
    class="helperText"
  />
</div>
`;

exports[`TextField It matches the snapshot 1`] = `
<div>
  <div
    class="textField"
  >
    <input
      class="inputField"
      type="text"
      value=""
    />
  </div>
  <p
    class="helperText"
  />
</div>
`;
